docs(js): Update span metrics instrumentation examples (#15320)
getsentry/sentry-docs: docs/platforms/javascript/common/tracing/span-metrics/examples.mdx

---
title: Example Instrumentation
description: >-
  Examples of using span metrics to debug performance issues and monitor
  application behavior across frontend and backend services.
sidebar_order: 10
og_image: /og-images/platforms-javascript-common-tracing-span-metrics-examples.png
---

<Alert>

The sample code contained within this page is for demonstration purposes only. It is not production-ready. Examples are structural and ultimately may not be for your specific language or framework.

</Alert>

> This guide provides practical examples of using span attributes and metrics to solve common monitoring and debugging challenges across your entire application stack. Each example demonstrates how to instrument both frontend and backend components, showing how they work together within a distributed trace to provide end-to-end visibility. You'll also find example repository code, walkthroughs and attributes to explore.

Reviewer: ```suggestion
This guide provides practical examples of using span attributes and metrics to solve common monitoring and debugging challenges across your entire application stack. Each example demonstrates how to instrument both frontend and backend components, showing how they work together within a distributed trace to provide end-to-end visibility. You'll also find example repository code, walkthroughs, and attributes to explore.
```

## E-Commerce Checkout Flow (React + Backend)

<Arcade src="https://demo.arcade.software/tGYGb3qewR6i4xjRt2Tg?embed" />

Example Repository: [Crash Commerce](https://github.com/getsentry/crash-commerce-tracing-sample)

> **Challenge:** Capture end-to-end checkout flow, understand average cart size and value, diagnose performance of payment providers across frontend, and server API.

Reviewer: ```suggestion
**The Challenge:** Capture end-to-end checkout flow, understand average cart size and value, diagnose performance of payment providers across the frontend and the server API.
```

            return result
          }
        )

        if (charge.status === 'failed' || !reserved) {
          span.setAttribute('payment.provider', charge.provider)
          span.setAttribute('payment.status', 'failed')
          span.setAttribute('inventory.reserved', reserved)
          res.status(402).json({ error: 'Payment failed' })
          return
        }

        const orderId = randomId()
        ORDERS.push({ id: orderId, totalMinor, items })

        // Set attributes before returning
        span.setAttribute('order.id', orderId)
        span.setAttribute('payment.provider', charge.provider)
        span.setAttribute('payment.status', 'success')
        span.setAttribute('inventory.reserved', reserved)

        res.json({ orderId, paymentProvider: charge.provider })
      } catch (err) {
        Sentry.captureException(err)
        res.status(500).json({ error: 'Internal error' })
      }
    }
  )
})
```

**How the trace works together:**
- UI span starts when checkout is selected -> Server Backend starts a span to continue the track when the server `/checkout` API is called. As payment processes, a payment span is started.

> - Attributes and Span metrics let you track more than just the latency of the request. Can track store busienss performances through `cart.item_count` and other `cart` attributes, and store reliabiliyt by checking error performance on `payment.provider` properties.

Reviewer 2: There's a typo in here. Should it be "Can track business performance through..."?

Reviewer 2: Also `reliabiliyt`

> - Attributes and Span metrics let you track more than just the latency of the request. Can track store busienss performances through `cart.item_count` and other `cart` attributes, and store reliabiliyt by checking error performance on `payment.provider` properties.

Reviewer: ```suggestion
- Attributes and Span metrics let you track more than just the latency of the request. You can also track the store's business performance through `cart.item_count` and other `cart` attributes, and store reliability by checking error performance on `payment.provider` properties.
```

What to monitor with span metrics:
- p95 span.duration of `op:ui.action` checkout by `cart.item_count` bucket.
- Error rate for `op:payment` by `payment.provider`.

## Media Upload with Background Processing (React + Express)

Example Repository: [SnapTrace](https://github.com/getsentry/snaptrace-tracing-example)

**Challenge:** Track user-perceived upload time, server-side validation, and async media processing (optimization, thumbnail generation) while maintaining trace continuity across async boundaries.

> **Solution:** Start a client span for the entire upload experience, create a backend span for upload validation, and a separate span for async media processing. Use rich attributes instead of excessive spans to capture processing details.

Reviewer: ```suggestion
**The Solution:** Start a client span for the entire upload experience, create a backend span for upload validation, and another span for async media processing. Use rich attributes instead of excessive spans to capture processing details.
```

      attributes: {
        'file.size_bytes': selectedFile.size,
        'file.mime_type': selectedFile.type,
      }
    },
    async (span) => {
      const uploadStartTime = Date.now();
      
      try {
        // Single API call to upload and start processing
        const uploadResponse = await fetch(`${API_BASE_URL}/api/upload`, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            fileName: selectedFile.name,
            fileType: selectedFile.type,
            fileSize: selectedFile.size
          })
        });

        if (!uploadResponse.ok) {
          throw new Error(`Upload failed: ${uploadResponse.statusText}`);
        }

        const uploadData = await uploadResponse.json();
        
        // Set success attributes
        span?.setAttribute('upload.success', true);
        span?.setAttribute('upload.duration_ms', Date.now() - uploadStartTime);
        span?.setAttribute('job.id', uploadData.jobId);
        
        // Update UI to show processing status
        updateUploadStatus(uploadData.jobId, 'processing');
        
      } catch (error) {

>         span?.setAttribute('upload.success', false);

Reviewer 2: We're missing a `span.setStatus` here.

        span?.setAttribute('upload.error', error instanceof Error ? error.message : 'Unknown error');
        setUploadStatus('error');
      }
    }
  );
};
```

Where to put this in your app:
- In the upload button click handler or form submit handler
- In drag-and-drop onDrop callback
- Auto-instrumentation will capture fetch spans; the explicit span adds business context

**Backend - Upload Validation and Queue Job**

<Alert>

> This example demonstrates proper queue instrumentation patterns. For more details on instrumenting queues, see the <PlatformLink to="/tracing/instrumentation/queues-module/">Queues Module documentation</PlatformLink>.

Reviewer: This link points to a 404:

https://sentry-docs-git-docs-span-metrics-examples.sentry.dev/platforms/javascript/tracing/instrumentation/queues-module/

                span?.setAttribute('processing.duration_ms', Date.now() - startTime);
                span?.setAttribute('result.size_saved_bytes', sizeSaved);
                span?.setAttribute('result.size_reduction_percent', 30);
                span?.setAttribute('result.status', 'success');
                
                // Update job status
                job.status = 'completed';
                
                // Mark parent span as successful
                parentSpan.setStatus({ code: 1, message: 'ok' });
                
              } catch (error) {
                span?.setAttribute('result.status', 'failed');
                span?.setAttribute('error.message', error instanceof Error ? error.message : 'Unknown error');
                parentSpan.setStatus({ code: 2, message: 'error' });
                Sentry.captureException(error);
              }
            }
          );
        }
      );
    }
  );
}
```

**How the trace works together:**
- Frontend span (`file.upload`) captures the entire user experience from file selection to server response.
- Backend producer span (`queue.publish`) tracks job enqueueing with proper queue attributes.

> - Consumer span (`queue.process`) continues the trace using `continueTrace()` with trace headers stored in the job.

Reviewer: ```suggestion
- The consumer span (`queue.process`) continues the trace using `continueTrace()` with trace headers stored in the job.
```

- Async processing runs independently with its own trace connected via queue instrumentation.
- Rich attributes on the consumer span capture all processing details without creating excessive child spans.
- This pattern populates Sentry's Queues insights page for monitoring queue performance.

What to monitor with span metrics:
- p95 upload duration by `file.size_bucket`.
- Processing success rate by `media.mime_type`.
- Average storage saved via `result.size_saved_bytes` where `result.status = success`.
- Queue latency via `messaging.message.receive.latency` to track processing delays.
- Job throughput via `op:queue.publish` and `op:queue.process` span counts.

## Search Autocomplete (debounced, cancellable, performance monitoring)

Example Repository: [NullFlix](https://github.com/getsentry/nullflix-tracing-example)

**Challenge:** Users type quickly in search; you need to debounce requests, cancel in-flight calls, handle errors gracefully, and monitor performance across different query types while keeping latency predictable.

> **Solution:** Start a client span for each debounced request, mark aborted requests, track search patterns, and on the server, instrument search performance with meaningful attributes.

Reviewer: ```suggestion
**The Solution:** Start a client span for each debounced request, mark aborted requests, track search patterns, and on the server, and instrument search performance with meaningful attributes.
```

        {
          signal: controller.signal,
          headers: { 'Content-Type': 'application/json' },
        }
      );

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        const errorMessage = errorData.error || `Search failed: ${response.status}`;
        throw new Error(errorMessage);
      }

      const data: SearchResponse = await response.json();
      
      span?.setAttribute('results.count', data.results.length);
      span?.setAttribute('results.has_results', data.results.length > 0);
      span?.setAttribute('http.response_size', JSON.stringify(data).length);
      span?.setStatus({ code: 1, message: 'ok' });
      
      return data;
    } catch (error) {
      if (error instanceof Error && error.name === 'AbortError') {
        span?.setAttribute('ui.aborted', true);
        span?.setStatus({ code: 2, message: 'cancelled' });
        // Don't re-throw AbortError to avoid sending it to Sentry
        return { results: [] };
      }
      
      span?.setStatus({ code: 2, message: error instanceof Error ? error.message : 'unknown error' });
      throw error;
    }
  }
);
```

Where to put this in your app:

> - In your search input component, triggered after debounce timeout

Reviewer: ```suggestion
- In your search input component, triggered after a debounce timeout
```

        }
        
        // Perform search
        const startSearch = Date.now();
        const results = await searchMovies(query);
        const searchDuration = Date.now() - startSearch;
        
        // Set span attributes
        span?.setAttribute('search.engine', 'elasticsearch');
        span?.setAttribute('search.mode', queryLength < 3 ? 'prefix' : 'fuzzy');
        span?.setAttribute('results.count', results.length);
        span?.setAttribute('query.length', queryLength);
        
        // Track slow searches
        if (searchDuration > 500) {
          span?.setAttribute('performance.slow', true);
          span?.setAttribute('search.duration_ms', searchDuration);
        }
        
        return res.json({ results });
      } catch (error: any) {
        span?.setStatus({ code: 2, message: error?.message || 'error' });
        span?.setAttribute('error.type', (error as any)?.constructor?.name || 'Error');
        
        Sentry.captureException(error);
        if (!res.headersSent) {
          return res.status(500).json({ error: 'Search failed' });
        }
      }
    }
  );
});
```

**How the trace works together:**

> - Client span starts when debounced search triggers -> tracks the full user-perceived latency.

Reviewer: ```suggestion
- The client span starts when the debounced search triggers, then tracks the full user-perceived latency.
```

- Aborted requests are marked with `ui.aborted=true` and short duration, showing wasted work.

> - Server span shows search performance characteristics: mode (prefix vs fuzzy), results count, and slow queries.

Reviewer: ```suggestion
- The server span shows search performance characteristics: mode (prefix vs fuzzy), results count, and slow queries.
```

What to monitor with span metrics:
- p95 duration of `op:search` grouped by `query.length`.
- Characteristics of slow searches via `op:search performance.slow:true`.
- Compare prefix vs fuzzy via `op:search` grouped by `search.mode`.
- Cancellation rate via `op:http.client ui.aborted:true`.
- Empty result rate via `op:http.client results.has_results:false`.
- Distribution of `http.response_size` for payload optimization.
- Error rate for `op:search` filtered by `status:error`.
- Backend abandonment via `op:search request.aborted:true`.

## Manual LLM Instrumentation (Custom AI Agent + Tool Calls)

Example Repository: [Customer Service Bot](https://github.com/getsentry/llm-manual-agent-monitoring-example)

> **Challenge:** You're building a custom AI agent that uses a proprietary LLM API (not OpenAI/Anthropic), performs multi-step reasoning with tool calls, and needs comprehensive monitoring to track token usage, tool performance, and agent effectiveness across the entire conversation flow.

Reviewer: ```suggestion
**The Challenge:** You're building a custom AI agent that uses a proprietary LLM API (not OpenAI/Anthropic), performs multi-step reasoning with tool calls, and needs comprehensive monitoring to track token usage, tool performance, and agent effectiveness across the entire conversation flow.
```

        });
        
      } catch (error) {
        agentSpan.setStatus({ code: 2, message: 'agent_invocation_failed' });
        agentSpan.setAttribute('error.type', error instanceof Error ? error.constructor.name : 'UnknownError');
        Sentry.captureException(error);
        res.status(500).json({ error: 'AI agent processing failed' });
      }
    }
  );
});

// Helper functions for tool execution
async function searchKnowledgeBase(query: string): Promise<string> {
  // Search company knowledge base - returns relevant policy info
  const results = [
    "Our return policy allows returns within 30 days of purchase.",
    "Refunds are processed within 5-7 business days after we receive the item.",
    "Items must be in original condition with tags attached.",
    "Free return shipping is provided for defective items."
  ];
  return results.join('\n');
}

async function synthesizeResponse(llmResponse: any, toolsUsed: string[]): Promise<any> {
  // Make final LLM call to synthesize tool results into response
  return {
    message: "Based on the information I found, here's your answer...",
    usage: { total_tokens: 150 }
  };
}
```

**How the trace works together:**

> - Frontend span (`gen_ai.invoke_agent`) captures the entire user interaction from message to response.

Reviewer: ```suggestion
- The frontend span (`gen_ai.invoke_agent`) captures the entire user interaction from message to response.
```

- Backend agent span continues the trace with the same operation and agent name for correlation.
- LLM spans (`gen_ai.chat`) track individual model calls with token usage and performance.

> - Tool execution spans (`gen_ai.execute_tool`) monitor each tool call with input/output and timing.

Reviewer: ```suggestion
- The tool execution spans (`gen_ai.execute_tool`) monitor each tool call with input/output and timing.
```
